### DISCUSSION TITLE
Issue with @request.data and existing fields api rules

### DISCUSSION THREAD

**Discussion by ghostdevv** [pocketbase/pocketbase#4688] [Q&A]
The context for my confusion is that I have two relations `poll` and `option` and I need to check that their id's are equal. For the create rule it's simple since we can use `poll.id = option.poll.id` as the `@request.data.*` and the existing fields.

My rule is failing, I presume because when I call update I am not always passing `poll` and `option` so the equality check of `@request.data.poll.id = @request.data.option.poll.id` fails. I wanted to add a rule that will use the `@request.data.*` or fallback to the existing field so the equality check will pass but couldn't figure it out

**Comment by ganigeorgiev** [comment]
The `@request.data.*` targets only the submitted request json/form data fields.

If you don't want to always submit the field values with the update you can try something like:

```js
(
    // both poll and option are NOT submitted so we don't have to check anything further
    (@request.data.poll:isset = false && @request.data.option:isset = false) ||
    // otherwise check the individual fields
    (
        // check option
        (
            // no option is submitted
            @request.data.option:isset = false ||
            // if there is no poll submitted, ensure that the submitted option's poll matches the existing one
            (@request.data.poll:isset = false && @request.data.option.poll = poll) ||
            // if there is poll submitted, ensure that the submitted option's poll matches the submitted one
            (@request.data.poll:isset = true && @request.data.option.poll = @request.data.poll)
        ) &&
        // check poll
        (
            // no poll is submitted
            @request.data.poll:isset = false ||
            // if there is no option submitted, ensure that the submitted poll matches the existing option's poll
            (@request.data.option:isset = false && @request.data.poll = option.poll) ||
            // if there is option submitted, ensure that the submitted poll matches the submitted option's poll
            (@request.data.option:isset = true && @request.data.poll = @request.data.option.poll)
        )
    )
)
```

But in general if possible I'd advice to consider normalizing your collections structure and try to have the relation only in one place to avoid "poll" and "options.poll" getting out of sync (eg. in case the related `option` record is updated). If that's not an option another possible alternative could be to ensure the correct "poll" and "options.poll" values are set programmatically via a before hook (_or even overwriting the submitted user value if you don't want to return errors_).

**Reply by ghostdevv** [reply]
> But in general if possible I'd advice to consider normalizing your collections structure and try to have the relation only in one place to avoid "poll" and "options.poll" getting out of sync (eg. in case the related option record is updated).

I initially had the db correctly normalised but I need to apply a unique constraint and there doesn't appear to be a way to do that across relations - unless you know of a better way with api rules to enforce this I'll have to stick with that

> If that's not an option another possible alternative could be to ensure the correct "poll" and "options.poll" values are set programmatically via a before hook (or even overwriting the submitted user value if you don't want to return errors).

I've considered setting this up this as we'll need to add some hooks in later anyway - but in this specific case I wanted to know if API rules are capable to save time

**Comment by ghostdevv** [comment]
@ganigeorgiev I think it's worth thinking about this situation more - take what I think is probably a common example where I have a user relation in my collection. If I have the create rule set to `@request.auth.id = user.id` perfect! However, if the update rule is set to that you have the same problem as I originally described but much worse. Someone could create a record then update it and change the user id and it's not immediately obvious to the dev why/that this can happen. It's compounded by the fact that if I changed the rule to `@request.auth.id = @request.user.id` it'll fail if user isn't set, so then we have to create a more complex and hard to grok rule to prevent against that.

**Reply by ganigeorgiev** [reply]
I don't think I can help you as I'm really confused and don't understand what actually you are trying to explain.
Again, `@request.data.*` targets only the submitted fields.
There is no such field as `@request.user.*`.
Also users cannot change their id.

**Reply by ganigeorgiev** [reply]
>  Someone could create a record then update it and change the user id and it's not immediately obvious to the dev why/that this can happen.

On second read, if by this you mean that someone could change the relation then that's OK and it is the developer responsibility to prevent it if they don't want this behavior.

But again, the `@request.data.*` field works as intended (_with the planned refactoring I'll consider renaming it to `@request.body.*` to be more specific what "data" is_).

**Reply by ghostdevv** [reply]
What I am really getting at is there should be a simpler way to write an api rule that uses either the existing field or the field that is about to prevent. It'll simplify authoring rules and it'll help prevent against mistakes like I outlined with the user one (and make the fix simpler).

> developer responsibility to prevent it if they don't want this behavior.

It's not easily clear what is happening, either when writing the rules or from the docs.
